src/protocol/payload/tx.rs

```diff
@@ -0,0 +1,868 @@
+use crate::protocol::payload::{read_n_bytes, Hash, VarInt};
+
+use std::io::{self, Cursor, Read, Write};
+
+/// A Zcash transaction ([spec](https://zips.z.cash/protocol/canopy.pdf#txnencodingandconsensus)).
+///
+/// Supports V1-V4, V5 isn't yet stable.
+#[derive(Debug, PartialEq)]
+pub enum Tx {
+    V1(TxV1),
+    V2(TxV2),
+    V3(TxV3),
+    V4(TxV4),
+    // Not yet stabilised.
+    V5,
+}
+
+impl Tx {
+    pub fn encode(&self, buffer: &mut Vec<u8>) -> io::Result<()> {
+        match self {
+            Tx::V1(tx) => {
+                // The overwintered flag is NOT set.
+                buffer.write_all(&1u32.to_le_bytes())?;
+                tx.encode(buffer)?;
+            }
+            Tx::V2(tx) => {
+                // The overwintered flag is NOT set.
+                buffer.write_all(&2u32.to_le_bytes())?;
+                tx.encode(buffer)?;
+            }
+            Tx::V3(tx) => {
+                // The overwintered flag IS set.
+                buffer.write_all(&(3u32 | 1 << 31).to_le_bytes())?;
+                tx.encode(buffer)?;
+            }
+            Tx::V4(tx) => {
+                // The overwintered flag IS set.
+                buffer.write_all(&(4u32 | 1 << 31).to_le_bytes())?;
+                tx.encode(buffer)?;
+            }
+            _ => unimplemented!(),
+        }
+
+        Ok(())
+    }
+
+    pub fn decode(bytes: &mut Cursor<&[u8]>) -> io::Result<Self> {
+        let (version, overwinter) = {
+            const LOW_31_BITS: u32 = (1 << 31) - 1;
+            let header = u32::from_le_bytes(read_n_bytes(bytes)?);
+
+            // Extract transaction version and check if overwinter flag is set.
+            (header & LOW_31_BITS, header >> 31 != 0)
+        };
+
+        let tx = match (version, overwinter) {
+            (1, false) => Self::V1(TxV1::decode(bytes)?),
+            (2, false) => Self::V2(TxV2::decode(bytes)?),
+            (3, true) => Self::V3(TxV3::decode(bytes)?),
+            (4, true) => Self::V4(TxV4::decode(bytes)?),
+            _ => unimplemented!(),
```

Comment:
Should we not return an Error indicating bad payload here instead? I know there is a version 5 on the way, but I wouldn't expect an `io::Result<()>` function to panic on a parsing error.